### DISCUSSION TITLE
Expansion Board Base for XIAO | Seeed Studio Wiki

### DISCUSSION THREAD

**Discussion by giscus** [Seeed-Studio/wiki-documents#1857] [Q&A]
# Expansion Board Base for XIAO | Seeed Studio Wiki

Seeed Studio XIAO Expansion board

https://wiki.seeedstudio.com/Seeeduino-XIAO-Expansion-Board/

**Comment by TimoteoMendes** [comment]
Hello
The FAQ says all pins 5v are from usb but that is not true as the schematics.
And find this because my board just stops working sundly without any reason as it was running program for couple of hours.

**Reply by TimoteoMendes** [reply]
I just hit enter...
Does anyone had experience this before?
Thanks

**Comment by lettcco** [comment]
Check the schematics...doesn't look like you can check the LIPO battery voltage?

**Comment by pop-aruaru** [comment]


**Reply by nfs0619** [reply]
Thank you for your question! In fact, on the XIAO board, both A3 and D3 point to the same GPIO, so you can use either one to connect your speaker. While the code references D3, you can also use A3 as they have the same functionality and will not affect the sound output. If you continue to encounter any issues, feel free to contact our technical support team at [techsupport@seeed.io](mailto:techsupport@seeed.io). We’ll be happy to assist you further!

**Comment by kzd76** [comment]
If you have issues like me with a xiao rp2040 + expansion board that the display is not working at all and the setup function takes minutes to execute, replace the 8x8 definition to 
U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(/* reset=*/ U8X8_PIN_NONE);

**Comment by microDevSys** [comment]
test with RP2040 in c++ under VScode work great : https://github.com/microDevSys/XIAO-RP2040-extension-board

**Reply by nfs0619** [reply]
Great to hear! Thanks for testing and sharing your experience.

**Comment by yu-zou** [comment]
I'm using nrf52840 sense and the expansion board. The example OLED code doesn't work at all.

**Reply by Lukilyy** [reply]
Hi, sorry for the inconvenience.

The OLED example in the Expansion Board Wiki was mainly written based on the XIAO SAMD21. When using XIAO nRF52840 Sense, the hardware connection is based on I2C, but the example code may need to be adjusted according to the nRF52840 Sense pin definition.

For XIAO nRF52840 Sense, the I2C pins are **D4 / SDA** and **D5 / SCL**. Please also make sure the correct board package and board model are selected in Arduino IDE.

You can try the following OLED test code:

```cpp
#include <Arduino.h>
#include <U8x8lib.h>
#include <Wire.h>

// XIAO nRF52840 Sense: SDA = D4, SCL = D5
U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(
  /* reset=*/ U8X8_PIN_NONE,
  /* clock=*/ D5,
  /* data=*/ D4
);

void setup(void) {
  u8x8.begin();
  u8x8.setFlipMode(1);   // Enable (1) or disable (0) 180 degree rotation of the display content
}

void loop(void) {
  u8x8.setFont(u8x8_font_chroma48medium8_r);
  u8x8.setCursor(0, 0);
  u8x8.print("Hello World!");
}
```

Please try adjusting the I2C pin settings accordingly. We look forward to your test results.

**Reply by yu-zou** [reply]
Thanks for the reply, but unfortunately it doesn't work. No output on OLED.

**Reply by grilliiiiii** [reply]
You can check the library you are using. If the wrong library is applied, it may result in the screen driver failing to start properly.

**Reply by FaiyuetCik** [reply]
Thank you for your feedback. We have noticed that you have tested the code using D4 and D5, but the OLED still does not display, so it is necessary to further distinguish between the software environment, I2C communication and hardware Question.

  We conducted tests using the following identical hardware:

  1. Seeed Studio XIAO nRF52840 Sense，SKU 102010469
  2. Expansion Board Base for XIAO，SKU 103030356
  3. Arduino IDE Board：Seeed XIAO nRF52840 Sense
  4. OLED: SSD1306, I2C address 0x3C
  5. RTC: PCF8563, I2C address 0x51
  6. U8g2: Version 2.36.19

  In our tests, the I2C Scanner was able to detect both the OLED and the RTC, and the OLED functioned properly for display. Therefore, this is not a widespread compatibility issue between these two products.

  Currently, the OLED examples on the Wiki are mainly written based on XIAO SAMD21, and the code using SCL and SDA is not applicable to all nRF52840 Arduino cores. Please follow the steps below
  Conduct a complete troubleshooting in accordance with the steps.

  1. Confirm the board selection in Arduino IDE

  Open in Arduino IDE:

  Tools > Board

  Please select the correct development board from the following options:

  Seeed XIAO nRF52840 Sense

  Please do not select the regular XIAO nRF52840 or other development boards compatible with the nRF52840.

  Please record the name and version of the installed board package at the same time, as this information will be required for subsequent troubleshooting.

  2. Confirm the U8g2 library

  Please search in the Arduino IDE Library Manager:

  U8g2 by olikraus

  Official Repository Link:

  https://github.com/olikraus/u8g2 (https://github.com/olikraus/u8g2)

  U8x8lib. h is already included in the U8g2 library. Please also record the version of U8g2 you have installed.

  3. Hardware Connection

  Disconnect the USB power supply, then perform the following operations:

  1. XIAO nRF52840 Sense and Expansion Board Base for XIAO.
  2. Check the installation orientation of XIAO.
  3. Make sure that the two rows of pins are fully inserted without misalignment or bending.
  4. Reconnect the USB.

  4. Run the I2C Scanner

  Please upload the scanner program below first, and do not run the OLED example beforehand:

 ```
 #include <Arduino.h>
  #include <Wire.h>

  void setup() {
    Serial.begin(115200);

    const unsigned long serialDeadline = millis() + 10000;

    while (!Serial && millis() < serialDeadline) {
      delay(10);
    }

    // XIAO nRF52840 Sense maps Wire to SDA D4 and SCL D5.
    Wire.begin();
    delay(500);

    Serial.println("XIAO Expansion Board I2C Scanner");
    Serial.println("Expected: OLED=0x3C, RTC=0x51");

    uint8_t count = 0;

    for (uint8_t address = 1; address < 127; address++) {
      Wire.beginTransmission(address);
      const uint8_t error = Wire.endTransmission();

      if (error == 0) {
        Serial.print("Found: 0x");

        if (address < 0x10) {
          Serial.print('0');
        }

        Serial.println(address, HEX);
        count++;
      }
    }

    Serial.print("Device count: ");
    Serial.println(count);
  }

  void loop() {
  }
```

  After uploading, open the Arduino IDE Serial Monitor and set the baud rate to:

` 115200`

  The expected normal result shall include:

  XIAO Expansion Board I2C Scanner
  Expected: OLED=0x3C, RTC=0x51
  Found: 0x3C
  Found: 0x51
  Device count: 2

  where 0x3C corresponds to the SSD1306 OLED, and 0x51 corresponds to the PCF8563 RTC.

  The scan results can be judged in the following ways:

  1. Both 0x3C and 0x51 are detected simultaneously, indicating that the I2C bus and the expansion board are basically connected properly, and the OLED test can proceed.
  2. If only 0x3C is detected, it indicates that the OLED communication is normal, and the OLED test can proceed. The RTC issue is a separate, independent problem.
  3. If only 0x51 is detected, it indicates that the I2C bus is normal, but the OLED does not respond. Please reinstall XIAO and scan again. If the result remains unchanged, it may be a hardware issue with the OLED or the expansion
     Display panel connection issue.

  4. Neither address is detected. Please check the installation orientation of XIAO, pin contacts, power supply, development board selection and board package.
  5. If any other addresses are detected, please provide the full scan output.

  5. Run OLED test

  Only after the I2C Scanner detects 0x3C, upload the following program:

 ```
 #include <Arduino.h>
  #include <U8x8lib.h>
  #include <Wire.h>

  // XIAO nRF52840 Sense uses SDA D4 and SCL D5.
  U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(
    U8X8_PIN_NONE,
    D5,
    D4
  );

  void setup() {
    Serial.begin(115200);

    const unsigned long serialDeadline = millis() + 10000;

    while (!Serial && millis() < serialDeadline) {
      delay(10);
    }

    // U8x8 uses the 8 bit form of the 7 bit I2C address.
    u8x8.setI2CAddress(0x3C * 2);
    u8x8.begin();
    u8x8. setPowerSave (0);
    u8x8.setFlipMode(1);
    u8x8.setFont(u8x8_font_chroma48medium8_r);

    u8x8.drawString(0, 0, "OLED TEST");
    u8x8.drawString(0, 2, "D4 SDA");
    u8x8.drawString(0, 3, "D5 SCL");

    Serial.println("OLED initialization completed");
  }

  void loop() {
  }
```

  Under normal conditions, the OLED should display:
 ```
  OLED TEST
  D4 SDA
  D5 SCL
```

  The serial monitor should display:

  `OLED initialization completed`

  If the I2C Scanner detects 0x3C and the serial port also shows that initialization is complete, yet the OLED remains entirely black, please provide the following information:

  1. The complete output of the I2C Scanner.
  2. Check whether Seeed XIAO nRF52840 Sense is selected in the Arduino IDE.
  3. Board package name and version.
  4. U8g2 library version.
  5. A clear photo of the XIAO mounted on the expansion board.
  6. Complete the compile and upload logs.

  With these results, we can further determine whether the issue lies with the Arduino core, the library version, hardware contact, or the OLED unit itself.
